src/Persistence_matrix/include/gudhi/matrix.h

```diff
@@ -0,0 +1,2129 @@
+
+  return matrix_.get_column_dimension(columnIndex);
+}
+
+template <class PersistenceMatrixOptions>
+template <typename Index_type>
+inline std::enable_if_t<std::is_integral_v<Index_type> > Matrix<PersistenceMatrixOptions>::add_to(
+    Index_type sourceColumnIndex, Index_type targetColumnIndex)
+{
+  matrix_.add_to(sourceColumnIndex, targetColumnIndex);
+}
+
+template <class PersistenceMatrixOptions>
+template <class Cell_range>
+inline std::enable_if_t<!std::is_integral_v<Cell_range> > Matrix<PersistenceMatrixOptions>::add_to(
+    const Cell_range& sourceColumn, index targetColumnIndex)
+{
+  static_assert(!isNonBasic,
+                "For boundary or chain matrices, only additions with columns inside the matrix is allowed to maintain "
+                "algebraic consistency.");
+
+  matrix_.add_to(sourceColumn, targetColumnIndex);
+}
+
+template <class PersistenceMatrixOptions>
+template <typename Index_type>
+inline std::enable_if_t<std::is_integral_v<Index_type> > Matrix<PersistenceMatrixOptions>::multiply_target_and_add_to(
+    Index_type sourceColumnIndex, int coefficient, Index_type targetColumnIndex) 
+{
+  if constexpr (PersistenceMatrixOptions::is_z2) {
+    // coef will be converted to bool, because of element_type
+    matrix_.multiply_target_and_add_to(sourceColumnIndex, coefficient % 2, targetColumnIndex);
+  } else {
+    matrix_.multiply_target_and_add_to(sourceColumnIndex, colSettings_->operators.get_value(coefficient), targetColumnIndex);
+  }
+}
+
+template <class PersistenceMatrixOptions>
+template <class Cell_range>
+inline std::enable_if_t<!std::is_integral_v<Cell_range> > Matrix<PersistenceMatrixOptions>::multiply_target_and_add_to(
+    const Cell_range& sourceColumn, int coefficient, index targetColumnIndex) 
+{
+  static_assert(!isNonBasic,
+                "For boundary or chain matrices, only additions with columns inside the matrix is allowed to maintain "
+                "algebraic consistency.");
+
+  if constexpr (PersistenceMatrixOptions::is_z2) {
+    // coef will be converted to bool, because of element_type
+    matrix_.multiply_target_and_add_to(sourceColumn, coefficient % 2, targetColumnIndex);
+  } else {
+    matrix_.multiply_target_and_add_to(sourceColumn, colSettings_->operators.get_value(coefficient), targetColumnIndex);
+  }
+}
+
+template <class PersistenceMatrixOptions>
+template <typename Index_type>
+inline std::enable_if_t<std::is_integral_v<Index_type> > Matrix<PersistenceMatrixOptions>::multiply_source_and_add_to(
+    int coefficient, Index_type sourceColumnIndex, Index_type targetColumnIndex) 
+{
+  if constexpr (PersistenceMatrixOptions::is_z2) {
+    // coef will be converted to bool, because of element_type
+    matrix_.multiply_source_and_add_to(coefficient % 2, sourceColumnIndex, targetColumnIndex);
+  } else {
+    matrix_.multiply_source_and_add_to(colSettings_->operators.get_value(coefficient), sourceColumnIndex, targetColumnIndex);
```

Comment:
Wouldn't it be convenient to give it one, or use a function that returns one, to avoid some `if constexpr`? The point of having a concept is that all models of the concept should provide the same interface. You will probably still keep some tests for performance reasons, but here there should be no performance impact of computing `%2` directly or inside a `get_value` function.

Reply:
There is no concept for `colSettings`, it is an `std::conditonal` between `Column_z2_settings` and `Column_zp_settings`. The idea was to hide later in them the different allocator strategies etc. for the different column types. But for now, I just used the opportunity to hide `operators` behind it to store or not store a pointer to it, even though it looks like an overkill in this state. I would wait until I know a bit more how these structures will look like before changing there interface too much.